staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -507,6 +524,15 @@ func (w *watchCache) WaitUntilFreshAndList(ctx context.Context, resourceVersion
 	return result, rv, index, err
 }
 
+func (w *watchCache) requestBookmarkIfNeeded(ctx context.Context, resourceVersion uint64) error {
+	w.RLock()
+	defer w.RUnlock()
+	if w.resourceVersion < resourceVersion {
+		return w.watch.RequestBookmark(ctx)
```

Comment:
Do we know how much time it will take?

How about some rare cases like broken connection, network partition etc?

It is more safe to take lock just for decision and then call the function without the lock

Reply:
right